File: leetcode/python/347_top_k_frequent_elements.py
import heapq
from typing import Counter


class Solution:
    def topKFrequent(self, nums: list[int], k: int) -> list[int]:
        counter = Counter(nums)
        heap = [(f, n) for n, f in counter.items()]
        heapq.heapify(heap)
        while len(heap) > k:
            heapq.heappop(heap)
        return [e[1] for e in heap]

s = Solution()
print(s.topKFrequent([1,2,3,1,2,1], 2))
print(s.topKFrequent([-1,-1], 1))
        

File: leetcode/rust/src/347_top_k_frequent_elements.rs
use std::collections::{HashMap, BinaryHeap};

pub fn top_k_frequent(nums: Vec<i32>, k: i32) -> Vec<i32> {
    let frequencies = nums
          .iter()
          .copied()
          .fold(HashMap::new(), |mut map, val|{
              map.entry(val)
                 .and_modify(|frq|*frq+=1)
                 .or_insert(1);
              map
          });

    let frequencies: Vec<(i32, i32)> = frequencies
        .into_iter()
        .map(|v| (v.1, v.0))
        .collect();
    let mut heap = BinaryHeap::from(frequencies);

    let mut res = vec![];
    for _ in 0..k {
        let (_, key) = heap.pop().unwrap();
        res.push(key);
    }
    res
}